dms/models/res_users.py

```diff
@@ -0,0 +1,116 @@
+            .directory_ids.filtered(lambda x: not x.is_hidden)
+        )
+        # User should be able to see all allowed directories, as well as the parent
+        # directories of the allowed directories, and that recursively
+        res_ids = set(my_directory.ids)
+
+        directory_from_attachment_ids = self._get_directory_from_attachment(operation)
+        res_ids |= set(directory_from_attachment_ids)
+        parent_ids = my_directory.parent_id.ids + directory_from_attachment_ids
+
+        while parent_ids:
+            res_ids |= set(parent_ids)
+            parent_ids = (
+                self.env["dms.directory"]
+                .search([("id", "in", parent_ids)])
+                .parent_id.ids
+            )
+
+        # Get the directories that the user has created
+        created_directories = self.env["dms.directory"].search(
+            [("create_uid", "=", self.env.uid)]
+        )
+        res_ids |= set(created_directories.ids)
+        return list(res_ids)
+
+    @api.model
+    def _has_all_access(self):
+        """
+        Check if user has access to all directories for the specified operation.
+
+        :return: The access group that has all access
+        :rtype: odoo.model.dms_access_group
+        """
+        if self.has_group("dms.group_dms_manager"):
+            return self.env["dms.access.group"].sudo().search([])
+        return self.env["dms.access.group"]
+
+    def _get_directory_from_attachment(self, operation):
+        """
+        Get the directory ids from the attachments that the user has access to. The
+        goal here is to allow the user to
+        access the directories based on the rule given in its profile settings. That
+        way, we are only using the standard
+        rule checking.
+
+        :param str operation: operation to check
+
+        :return: The directory ids that the user has access to
+        :rtype: list[int]
+        """
+
+        # get the directories based on a record
+        attachment_directories = self.env["dms.directory"].search(
+            [
+                ("res_id", "not in", [False, 0]),
+                ("res_model", "!=", False),
+            ]
+        )
+        if not attachment_directories:
+            return []
+
+        authorized_directory_ids = []
+        for directory in attachment_directories:
+            # use of .get because on the first startup of Odoo, all the modules might
+            # not be loaded yet
+            model = self.env.get(directory.res_model)
+            if model is not None:
+                record = model.browse(directory.res_id)
+                # standard way to check access rights
+                invalid = record - record.with_user(
+                    record.env.user
+                )._filter_access_rules_python(operation.replace("perm_", ""))
+                if not invalid.exists():
+                    authorized_directory_ids.append(directory.id)
+
+        return authorized_directory_ids
```

Comment:
IMO this change is not necessary, it is confusing and incorrect (I see `.sudo()` is applied at the beginning, which will cause all records queried to have the 'wrong permissions'). can we leave it as is in v16?

Reply:
This is still pending

Reply:
Sorry, I had missed this.

If something doesn't work in v16 it should be resolved in v16 first.
It is intentional that the access rules compare with `user.id`, the goal is that the `_compute_permissions()` method set everything correctly (https://github.com/OCA/dms/pull/326).

Checking the use case you indicate in v16 in runboat, it works correctly.

Use case:

Marc Demo: Permissions Document > User and Sale > User: Own Documents Only (**no more**)

![storage](https://github.com/OCA/dms/assets/4117568/00ddaab1-1d7b-41e6-a1be-afdbd2341931)
![carpetas](https://github.com/OCA/dms/assets/4117568/05e7f5c9-acbd-42da-9211-b55ce2bd5f6d)

S00007 > Order is assigned to Mitchell Admin
S00016 > Order is assigned to Marc Demo

If we log in with Marc Demo, he can only see the folder **S00016**.
![ejemplo-marc-demo](https://github.com/OCA/dms/assets/4117568/52c8457a-0606-4f4f-907b-c9cd2bfdfda4)

Having cleared all this up, if it works correctly in v16, it should also work in v17 without all these changes (similar to v16).

Reply:
Indeed, I don't know what I did, but following your instructions everything works. I rolled back all the changes I made about it.